core/Application/Environment.php

```diff
@@ -0,0 +1,176 @@
+    {
+        $this->environment = $environment;
+        $this->definitions = $definitions;
+    }
+
+    /**
+     * Initializes the kernel globals and DI container.
+     */
+    public function init()
+    {
+        $this->container = $this->createContainer();
+
+        StaticContainer::set($this->container);
+
+        $this->validateEnvironment();
+
+        Piwik::postEvent('Environment.bootstrapped'); // this event should be removed eventually
+    }
+
+    /**
+     * Returns the DI container. All Piwik objects for a specific Piwik instance should be stored
+     * in this container.
+     *
+     * @return Container
+     */
+    public function getContainer()
+    {
+        return $this->container;
+    }
+
+    /**
+     * Returns the Piwik object for this Environment's Piwik instance by ID.
+     *
+     * @param string $id
+     * @return mixed
+     * @throws \DI\NotFoundException
+     */
+    public function get($id)
+    {
+        return $this->container->get($id);
+    }
+
+    /**
+     * @link http://php-di.org/doc/container-configuration.html
+     */
+    private function createContainer()
+    {
+        $pluginList = $this->getPluginListCached();
+        $settings = $this->getGlobalSettingsCached();
+        $definitions = array_merge(StaticContainer::getDefinitions(), $this->definitions);
+
+        $containerFactory = new ContainerFactory($pluginList, $settings, $this->environment, $definitions);
+        return $containerFactory->create();
+    }
+
+    protected function getGlobalSettingsCached()
+    {
+        if ($this->globalSettingsProvider === null) {
+            $this->globalSettingsProvider = $this->getGlobalSettings();
+        }
+        return $this->globalSettingsProvider;
+    }
+
+    protected function getPluginListCached()
+    {
+        if ($this->pluginList === null) {
+            $this->pluginList = $this->getPluginList();
+        }
+        return $this->pluginList;
+    }
+
+    /**
+     * Returns the kernel global GlobalSettingsProvider object. Derived classes can override this method
+     * to provide a different implementation.
+     *
+     * @return null|GlobalSettingsProvider
+     */
+    protected function getGlobalSettings()
+    {
+        // TODO: need to be able to set path global/local/etc. which is in DI... for now works because TestingEnvironment creates
+        //       singleton instance before this method.
+        return IniSettingsProvider::getSingletonInstance();
+    }
+
+    /**
+     * Returns the kernel global PluginList object. Derived classes can override this method to
+     * provide a different implementation.
+     *
+     * @return PluginList
+     */
+    protected function getPluginList()
+    {
+        // TODO: in tracker should only load tracker plugins. can't do properly until tracker entrypoint is encapsulated.
+        return new IniPluginList($this->getGlobalSettingsCached());
```

Comment:
No, it will apply DI config for all plugins in tracker now though. I plan on creating a environment subclass for the Tracker, which will be used by the encapsulated tracker application.